src/Orleans/Runtime/CallbackData.cs

```diff
@@ -207,5 +205,124 @@ public TimeSpan RequestedTimeout()
+                    }
+
+                    queue.Enqueue(callback);
+                });
+            }
+
+
+            public void Register(CallbackData data)
+            {
+                queueChecker.Register(data);
+            }
+
+            public void RegisterResendable(CallbackData data)
+            {
+                if (resendableQueueChecker == null)
+                {
+                    throw new InvalidOperationException("Can't enqueue resendable callback when initialized with resendOnCallback = false");
+                }
+
+                resendableQueueChecker.Register(data);
+            }
+
+            public void Dispose()
+            {
+                queueChecker.Dispose();
+                resendableQueueChecker?.Dispose();
+            }
+
+            private class QueueChecker : IDisposable
+            {
+                private readonly Queue<CallbackData> callbacks = new Queue<CallbackData>();
+                private readonly FastLock queueLock = new FastLock();
+                private readonly SafeTimer queueChecker;
+                private readonly Action<Queue<CallbackData>, CallbackData> _onTimeout;
+
+                public QueueChecker(TimeSpan timerPeriod, Action<Queue<CallbackData>, CallbackData> onTimeout)
+                {
+                    _onTimeout = @onTimeout;
+                    queueChecker = new SafeTimer(state =>
+                    {
+                        CheckQueue();
+                    }, null, timerPeriod, timerPeriod);
+                }
+
+                public void Register(CallbackData data)
+                {
+                    queueLock.Take();
+                    callbacks.Enqueue(data);
+                    queueLock.Release();
+                }
+
+                // Crawls through the callbacks and timeouts expired ones
+                private void CheckQueue()
+                {
+                    var now = DateTime.UtcNow;
+                    while (true)
+                    {
+                        queueLock.Take();
+                        int maxItemsPerLock = 7;
+                        for (int i = 0; i < maxItemsPerLock; i++)
+                        {
+                            if (callbacks.Count == 0)
+                            {
+                                queueLock.Release();
+                                return;
+                            }
+
+                            var callback = callbacks.Peek();
+                            if (callback.alreadyFired)
+                            {
+                                callbacks.Dequeue();
+                                continue;
+                            }
+
+                            if (callback.dueTime < now)
+                            {
+                                callbacks.Dequeue();
+                                callback.OnTimeout();
+                                _onTimeout(callbacks, callback);
```

Comment:
is it guaranteed that these callbacks are extremely fast? otherwise the contention in queueLock might cause the CPU to spin very aggressively